crates/services/consensus_module/poa/src/ports.rs

```diff
@@ -142,3 +142,42 @@ impl PredefinedBlocks for InMemoryPredefinedBlocks {
 pub trait GetTime: Send + Sync {
     fn now(&self) -> Tai64;
 }
+
+pub trait TriggerBlockProduction: Send + Sync {
+    fn wait_for_trigger(&self) -> impl core::future::Future<Output = ()> + Send;
+}
+
+pub(crate) struct BlockProductionTrigger<TBP> {
+    notifier: TBP,
+    notified: std::sync::Arc<std::sync::atomic::AtomicBool>,
+}
+
+impl<TBP> TriggerBlockProduction for BlockProductionTrigger<TBP>
+where
+    TBP: TriggerBlockProduction,
+{
+    /// Cache the notification to avoid waiting for the trigger multiple times.
+    async fn wait_for_trigger(&self) {
+        if self.notified.load(std::sync::atomic::Ordering::Acquire) {
+            return;
+        }
+        self.notifier.wait_for_trigger().await;
+        // this is the first and only time we are notified
+        tracing::info!("Block Production has been triggered.");
+
+        self.notified
+            .store(true, std::sync::atomic::Ordering::Release);
+    }
+}
+
+impl<TBP> BlockProductionTrigger<TBP>
+where
+    TBP: TriggerBlockProduction,
+{
+    pub(crate) fn new(notifier: TBP) -> Self {
+        Self {
+            notifier,
+            notified: std::sync::Arc::new(std::sync::atomic::AtomicBool::new(false)),
+        }
+    }
+}
```

Comment:
This looks wrong to me.

1. We should only await the notifier once when the service is started, so in the expected case this is dead code.
2. If we in any place chose to restart the service, wouldn't we like to retain the same control in that scenario that this service only is started when the dependencies has started?

I tried removing this and all tests in the PoA module still pass for me. I think this type should be removed.

Reply:
the notion of a service being "started" doesn't really exist, it's just initialization and the run-loop.

<img width="961" alt="image" src="https://github.com/user-attachments/assets/ff44b17c-38c0-40de-adaa-a8472e1e2d81" />
from https://docs.rs/tokio/latest/tokio/sync/struct.Notify.html

the tests all probably just complete in a single loop, hence pass. the integ tests would fail. we don't wait to wait every single time in the run loop, which is why we cache the result from the first notify.

Reply:
> the notion of a service being "started" doesn't really exist, it's just initialization and the run-loop.

I'd consider the act of starting the run loop either explicitly or implicitly through `service_runner.start_and_await()` starting the service. But you're right that the service runner may call this multiple times, so I can see the point of caching the value now.